File: WINGS_AI_Bootcamp/Python/Odometer.py
def is_valid(n: int) -> bool:
    if n < 10:
        return True
    elif (n // 10) % 10 < n % 10:
        return is_valid(n // 10)
    else:
        return False

def get_limits(n: int) -> tuple[int, int]:
    LIMIT = "123456789"
    size = len(str(n))
    return int(LIMIT[:size]), int(LIMIT[-size:])

def next_reading(reading: int) -> int:
    start, limit = get_limits(reading)
    if reading == limit:
        return start
    reading += 1
    while not is_valid(reading):
        reading += 1
        if reading > limit:
            reading = start
    return reading

def next_reading_k(reading: int, step: int = 1) -> int:
    for _ in range(step):
        reading = next_reading(reading)
    return reading

def prev_reading(reading: int) -> int:
    start, limit = get_limits(reading)
    if reading == start:
        return limit
    reading -= 1
    while not is_valid(reading):
        reading -= 1
        if reading < start:
            reading = limit
    return reading

def prev_reading_k(reading: int, step: int = 1) -> int:
    for _ in range(step):
        reading = prev_reading(reading)
    return reading

def distance(a_reading: int, b_reading: int) -> int:
    if len(str(a_reading)) != len(str(b_reading)):
        return -1
    diff = 0
    current = a_reading
    while current != b_reading:
        current = next_reading(current)
        diff += 1
        if diff > 10**len(str(a_reading)):  # avoid infinite loop
            return -1
    return diff

# Examples
print(next_reading(2468))        # next reading after 2468
print(next_reading_k(2468, 3))   # 3 readings after 2468
print(prev_reading(2468))        # previous reading before 2468
print(prev_reading_k(2468, 3))   # 3 readings before 2468
print(distance(2468, 2469))      # distance between readings

File: WINGS_AI_Bootcamp/Elixir/Odometer.ex
defmodule Odometer do
  @limit "123456789"

  def is_valid(n) when n < 10, do: true

  def is_valid(n) do
    if rem(div(n, 10), 10) < rem(n, 10) do
      is_valid(div(n, 10))
    else
      false
    end
  end

  def get_limits(n) do
    size = String.length(Integer.to_string(n))
    start = String.slice(@limit, 0, size) |> String.to_integer()
    limit = String.slice(@limit, -size, size) |> String.to_integer()
    {start, limit}
  end

  def next_reading(reading) do
    {start, limit} = get_limits(reading)

    cond do
      reading == limit -> start
      true ->
        next = reading + 1
        if is_valid(next), do: next, else: next_reading(next)
    end
  end

  def next_reading_k(reading, step \\ 1) do
    Enum.reduce(1..step, reading, fn _, acc -> next_reading(acc) end)
  end

  def prev_reading(reading) do
    {start, limit} = get_limits(reading)

    cond do
      reading == start -> limit
      true ->
        prev = reading - 1
        if is_valid(prev), do: prev, else: prev_reading(prev)
    end
  end

  def prev_reading_k(reading, step \\ 1) do
    Enum.reduce(1..step, reading, fn _, acc -> prev_reading(acc) end)
  end

  def distance(a_reading, b_reading) do
    a_str = Integer.to_string(a_reading)
    b_str = Integer.to_string(b_reading)

    if String.length(a_str) != String.length(b_str) do
      -1
    else
      do_distance(a_reading, b_reading, 0, String.length(a_str) * 10)
    end
  end

  defp do_distance(_curr, _target, count, max) when count > max, do: -1
  defp do_distance(curr, target, count, _max) when curr == target, do: count

  defp do_distance(curr, target, count, max) do
    do_distance(next_reading(curr), target, count + 1, max)
  end
end

IO.inspect(Odometer.next_reading(2468))
IO.inspect(Odometer.next_reading_k(2468, 3))
IO.inspect(Odometer.prev_reading(2468))
IO.inspect(Odometer.prev_reading_k(2468, 3))
IO.inspect(Odometer.distance(2468, 2469))

